KAFKA-17377: Consider using defaultApiTimeoutMs in AsyncKafkaConsumer#unsubscribe

FrankYang0529 opened the pull request:
Followup from https://github.com/apache/kafka/pull/16673#discussion_r1717009724

The `AsyncKafkaConsumer#unsubscribe` uses `Long.MAX_VALUE` to calculate deadline. However, most of `AsyncKafkaConsumer` operations use [default.api.timeout.ms](https://kafka.apache.org/documentation/#consumerconfigs_default.api.timeout.ms) if users don't specify a timeout. In [design document](https://cwiki.apache.org/confluence/display/KAFKA/Java+client+Consumer+timeouts#JavaclientConsumertimeouts-DefaultTimeouts), it also mentions that using `default.api.timeout.ms` as default timeout.

The `AsyncKafkaConsumer#unsubscribe` uses `processBackgroundEvents` to wait for callbacks and a response to the leave group request, so this PR ensures that the api call is bounded by the `default.api.timeout.ms`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kirktrue reviewed (approved):
Thanks for the PR @FrankYang0529!

This makes sense to me.

Any concerns, @lianetm or @philipnee?

kirktrue commented:
Apologies @FrankYang0529! I failed to notice this was in draft form. It still looks good, though :)

FrankYang0529 committed:
KAFKA-17377: Consider using defaultApiTimeoutMs in AsyncKafkaConsumer#unsubscribe

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 commented:
Hi @kirktrue, thanks for review. I converted to draft to see CI result first. The previous result looks good. Failed cases can't be reproduced on my laptop. I also added related unit test for it. Could you review it again? Thanks.

kirktrue reviewed (approved):
Good call out on the unit test. Thanks!

lianetm reviewed (approved):
Hey @FrankYang0529 , thanks for the patch, LGTM. 

Could you just update the PR description maybe, to mention why this change is relevant (the unsubscribe call blocks waiting for callbacks and a response to the leave group request, so this PR ensures that the api call is bounded by the default api timeout). Thanks!

FrankYang0529 commented:
Hi @lianetm, thanks for the suggestion and review. Updated it.

lianetm commented:
Build completed with 8 unrelated test failures. Thanks @FrankYang0529 !
